### PR TITLE
User-Level Sampling (ULS) for Transformers (#88) Split PR 1

### PR DESCRIPTION
**Pull request by debanganghosh08**
This PR adds a standalone example implementing User-Level Sampling (ULS) for Transformers, as described in Charles et al. (2024), "Fine-Tuning Large Language Models with User-Level Differential Privacy". [Reference](https://arxiv.org/abs/2407.07737) 

**Correctness Verification**: To verify the implementation of the ULS mechanism, I implemented manual logging to ensure that intra-user averaging occurs strictly before the user-level sensitivity clip.

**Averaging Logic**: The script logs the number of sequences per user (e.g., User A: 4 sequences) being collapsed into a single gradient.

**Sensitivity Bounding**: I've added logs for the avg_grad_norm (pre-clip) to verify that the user_clip_norm is correctly applied to the aggregated user update rather than individual records.

This implementation uses a _manual loop_ to handle the varying amounts of _data per user_, ensuring stability in JAX without the need for c_omplex masking of variable-length batches_.

### CODE REVIEW

File: examples/user_level_transformer_example.py [hunk 1]
```diff
@@ -0,0 +1,311 @@
...
+with User-Level Differential Privacy" (https://arxiv.org/abs/2407.07737)
+
+--- Implementation Notes ---
+Standard DP-SGD clips every example. ULS (User-Level Sampling) is different:
+it averages all gradients from a single user into a 'user-update' BEFORE clipping.
+As noted in Sec 4.2 of the paper, this is much more efficient for LLMs because
+it improves the signal-to-noise ratio by reducing the magnitude of the update
+before we hit it with the sensitivity clip.
+
+We use a manual for-loop for the user-aggregation because users often have
```
Comment (ryan112358) [inline examples/user_level_transformer_example.py]: Please implement this in terms of the following components:

1. jax_privacy.batch_selection.UserSelectionStrategy
2. jax_privacy.clipped_grad(keep_batch_dim=False)

File: examples/user_level_transformer_example.py [hunk 2]
```diff
@@ -0,0 +1,311 @@
...
+padding and masking issues that can arise with JAX's vmap.
+"""
+
+import jax
+import jax.numpy as jnp
+import numpy as np
+from jax import random, tree_util
+from jax.flatten_util import ravel_pytree
+import optax
+import tensorflow as tf
```
Comment (ryan112358) [inline examples/user_level_transformer_example.py]: Can we eliminate the tensorflow dependency?

File: examples/user_level_transformer_example.py [hunk 3]
```diff
@@ -0,0 +1,311 @@
...
+
+        # Final projection
+        logits = x @ params["output_projection"]
+        return logits
+
+
+def loss_fn(params, batch):
+    """Cross-entropy loss for language modeling."""
+    inputs, targets = batch[:-1], batch[1:]
+    logits = model.apply_fn(params, inputs)
```
Comment (ryan112358) [inline examples/user_level_transformer_example.py]: Where does model come from?

File: examples/user_level_transformer_example.py [hunk 4]
```diff
@@ -0,0 +1,311 @@
...
+    log_probs = jax.nn.log_softmax(logits)
+    return -jnp.mean(jnp.take_along_axis(log_probs, targets[:, None], axis=-1))
+
+
+def user_level_grad_fn(params, user_data, user_clip_norm):
```
Comment (ryan112358) [inline examples/user_level_transformer_example.py]: Use jax_privacy.clipped_grad(keep_batch_dim=False) here

File: examples/user_level_transformer_example.py [hunk 5]
```diff
@@ -0,0 +1,311 @@
...
+    updates, opt_state = optimizer.update(noisy_grads, opt_state, params)
+    params = optax.apply_updates(params, updates)
+
+    return params, opt_state, noise_state
+
+
+# --- Configuration ---
+config = {
+    "num_steps": 10,
+    "batch_size": 2,  # Number of users per batch
```
Comment (ryan112358) [inline examples/user_level_transformer_example.py]: Let's call batch_size --> users_per_batch and add a new config option for examples_per_user

File: examples/user_level_transformer_example.py [hunk 6]
```diff
@@ -0,0 +1,311 @@
...
+)
+
+opt_state = optimizer.init(params)
+noise_state = privatizer.init(params)
+
+# --- Training Loop ---
+print("Starting user-level DP training...")
+for step in range(config["num_steps"]):
+    # Sample a batch of users for this step
+    sampled_user_ids = np.random.choice(
```
Comment (ryan112358) [inline examples/user_level_transformer_example.py]: Let's use jax_privacy.batch_selection, sampling users with poisson sampling and selecting examples by using the UserSelectionStrategy

File: examples/user_level_transformer_example.py [hunk 7]
```diff
@@ -0,0 +1,311 @@
...
+        self.id_to_char = {i: c for i, c in enumerate(self.vocab)}
+
+    def encode(self, text):
+        return np.array([self.char_to_id[c] for c in text], dtype=np.int32)
+
+    def decode(self, ids):
+        return "".join([self.id_to_char[i] for i in ids])
+
+
+class TransformerDecoder:
```
Comment (ryan112358) [inline examples/user_level_transformer_example.py]: As we begin to add more example scripts, there is some redundancy in the model definition and data loading.  I wonder if you can have a single standard model definition (maybe written in flax nnx or equinox), and then just import it here and in the other scripts?

File: examples/dp_sgd_transformer_nnx.py [hunk 8]
```diff
@@ -0,0 +1,369 @@
...
+  # and attention expectations which look for a 4D mask derived from
+  # a batched input. We index [0] on the output to remove it.
+  # x indices: [seq_len] -> [1, seq_len]
+  logits = m(x[None, :])[0]
+
+
+  return optax.softmax_cross_entropy_with_integer_labels(logits, y).mean()
+
+
+def main():
```
Comment (Neerajpathak07) [inline examples/dp_sgd_transformer_nnx.py]:
```suggestion
def main(argv: Sequence[str]) -> None:
  if len(argv) > 1:
    raise app.UsageError('Too many command-line arguments.')
```
feel free to add :-
```python
from typing import Sequence
```
if you think this suggestion makes sense!!!

File: examples/dp_sgd_transformer_nnx.py [hunk 9]
```diff
@@ -0,0 +1,369 @@
...
+  # Configure DP Gradient Clipping
+  grad_fn = clipped_grad(
+      functools.partial(pure_loss_fn, graphdef=graphdef, other=other),
+      l2_clip_norm=CLIP_NORM,
+      batch_argnums=(1, 2),  # x and y are batched
+      keep_batch_dim=False,  # Process per-example
+      return_values=True     # Return loss values for logging
+  )
+
+  privatizer = noise_addition.gaussian_privatizer(
```
Comment (Neerajpathak07) [inline examples/dp_sgd_transformer_nnx.py]: I believe you'll have to update this as `noise_addition` should be wired through `plan` which in turn has to be configured first.
Ref:- https://github.com/google-deepmind/jax_privacy/blob/84daf71e6e7070edd3e1e16245723e8a60dcd062/examples/dp_logistic_regression.py#L88

File: examples/user_level_transformer_example.py [hunk 10]
```diff
@@ -0,0 +1,182 @@
...
+    'users_per_batch', 4, 'Number of users to select in each batch.'
+)
+_EXAMPLES_PER_USER = flags.DEFINE_integer(
+    'examples_per_user', 2, 'Number of examples to select for each user.'
+)
+_STEPS = flags.DEFINE_integer('steps', 10, 'Number of training steps.')
+_L2_CLIP_NORM = flags.DEFINE_float(
+    'l2_clip_norm', 1.0, 'L2 clipping norm for gradients.'
+)
+_LEARNING_RATE = flags.DEFINE_float('learning_rate', 1e-3, 'Learning rate.')
```
Comment (Neerajpathak07) [inline examples/user_level_transformer_example.py]: Hey, @debanganghosh08 instead of flagging it like this here let's create hyperparameters for this and provide the apt value.
something similar has also been done in the upstream versions.
Ref:- https://github.com/google-deepmind/jax_privacy/blob/84daf71e6e7070edd3e1e16245723e8a60dcd062/examples/dp_logistic_regression.py#L28

File: examples/user_level_transformer_example.py [hunk 11]
```diff
@@ -0,0 +1,182 @@
...
+  num_users = 20
+  num_examples_per_user = np.random.randint(1, 10, size=num_users)
+
+  data, labels, user_ids = get_synthetic_data(
+      num_users=num_users,
+      num_examples_per_user=num_examples_per_user,
+      seq_len=seq_len,
+      vocab_size=vocab_size,
+  )
+
```
Comment (Neerajpathak07) [inline examples/user_level_transformer_example.py:114]: @ryan112358 Correct me if I am wrong but one thing that I have learned is that while applying poisson sampling for any example the best practice is to jot it down from `execution_plan.BandMFExecutionPlanConfig` and create a `plan` var to clipp the grad function.
Maybe we can do something similar here?

File: examples/user_level_transformer_example.py [hunk 12]
```diff
@@ -0,0 +1,182 @@
...
+  model = TransformerDecoder(
+      vocab_size=vocab_size,
+      embed_dim=embed_dim,
+      num_heads=num_heads,
+      ff_dim=ff_dim,
+  )
+  params = model.init(
+      jax.random.key(0), jnp.zeros((1, seq_len), dtype=jnp.int32), train=False
+  )['params']
+  optimizer = optax.adam(_LEARNING_RATE.value)
```
Comment (Neerajpathak07) [inline examples/user_level_transformer_example.py]:
```suggestion
  optimizer = optax.sgd(LEARNING_RATE)
```
where `LEARNING_RATE` will be the hyperparameter value hardcoded at top.

File: examples/user_level_transformer_example.py [hunk 13]
```diff
@@ -0,0 +1,182 @@
...
+  opt_state = optimizer.init(params)
+
+  # 2. Batch Selection
+  sampling_prob = _USERS_PER_BATCH.value / num_users
```
Comment (Neerajpathak07) [inline examples/user_level_transformer_example.py]:
```suggestion
  sampling_prob = EXPECTED_BATCH_SIZE / USERS
```
maybe we can update this for `EXPECTED_BATCH_SIZE` and add it as an argument for `config`
ref:- https://github.com/google-deepmind/jax_privacy/blob/84daf71e6e7070edd3e1e16245723e8a60dcd062/examples/dp_logistic_regression.py#L75

File: examples/user_level_transformer_example.py [hunk 14]
```diff
@@ -0,0 +1,182 @@
...
+  @jax.jit
+  def train_step(params, opt_state, batch_data, batch_labels):
+    grads = grad_fn(params, batch_data, batch_labels)
+    updates, opt_state = optimizer.update(grads, opt_state, params)
+    params = optax.apply_updates(params, updates)
+    return params, opt_state
+
+  # 4. Training Loop
+  batch_iterator = user_strategy.batch_iterator(user_ids, rng=0)
+  for step, user_batch_indices in enumerate(batch_iterator):
```
Comment (Neerajpathak07) [inline examples/user_level_transformer_example.py:187]:
```suggestion
  for step, user_batch_indices in enumerate(plan.batch_selection_strategy.batch_iterator(USERS)):
```
for poisson sampling this should come from `batch_selection_strategy`

File: examples/dp_sgd_transformer_nnx.py (hunk 8 above)
Comment (debanganghosh08) [reply examples/dp_sgd_transformer_nnx.py]: Thanks for the suggestion, Neeraj! Adding the Sequence[str] type hint and the UsageError check makes the entry point much more robust. I'll include this in the next push.

File: examples/user_level_transformer_example.py (hunk 10 above)
Comment (debanganghosh08) [reply examples/user_level_transformer_example.py]: Good point on the configuration style. Moving from flags to hardcoded constants at the top aligns better with the dp_logistic_regression template. I'll refactor the config block to use constants like USERS and LEARNING_RATE.

File: examples/user_level_transformer_example.py (hunk 12 above)
Comment (debanganghosh08) [reply examples/user_level_transformer_example.py]: I used adam here because it's the standard for Transformer convergence, but I'm happy to switch to sgd if we want to keep the example as simple as possible. @ryan112358, do you have a preference for the optimizer type in this example?

File: examples/user_level_transformer_example.py (hunk 13 above)
Comment (debanganghosh08) [reply examples/user_level_transformer_example.py]: Regarding the execution_plan and BandMFExecutionPlanConfig: I see how this is used in the upstream logistic regression example. However, for this Transformer example, I implemented the loop manually to make the interaction between UserSelectionStrategy and clipped_grad explicit for users. Do you think moving to the plan abstraction provides enough benefit to outweigh the loss of visibility into the DP mechanics for this specific use case?

File: examples/user_level_transformer_example.py (hunk 12 above)
Comment (Neerajpathak07) [reply examples/user_level_transformer_example.py]: here the best thing to do is compare what both of them do.

`Adam`-> Adam is an SGD variant with gradient scaling adaptation. The scaling used for each parameter is computed from estimates of first and second-order moments of the gradients (using suitable exponential moving averages).

`sgd` -> A canonical Stochastic Gradient Descent optimizer.
This implements stochastic gradient descent. It also includes support for momentum, and Nesterov acceleration, as these are standard practice when using stochastic gradient descent to train deep neural networks.

this should give you a broad idea of what would be a better fit here.

File: examples/user_level_transformer_example.py (hunk 13 above)
Comment (Neerajpathak07) [reply examples/user_level_transformer_example.py]: Yes, apparently we are not just clipping the grads just for users, we also want this for `sampling_prob` and other such parameters. Having them consisted at `clipped_grad` and wiring poisson sampling is the Jax_privacy way of protecting the dp mechanics.

File: examples/user_level_transformer_example.py (hunk 11 above)
Comment (ryan112358) [reply examples/user_level_transformer_example.py:114]: The ExecutionPlanConfig can be used or ignored, up to you.  It's kind of in an experimental state right now and the API might change in the future

File: examples/user_level_transformer_example.py (hunk 12 above)
Comment (ryan112358) [reply examples/user_level_transformer_example.py]: +1 to keeping adam here (or maybe switching to adamw).  Although might be worth trying out both as Neeraj suggests and seeing which one works better

File: examples/dp_sgd_transformer_nnx.py [hunk 15]
```diff
@@ -0,0 +1,390 @@
...
+  # graph and other state.
+  graphdef, params, other = nnx.split(model, nnx.Param, ...)
+  opt_state = optimizer.init(params)
+
+  # Configure DP Gradient Clipping
+  grad_fn = clipped_grad(
+      functools.partial(pure_loss_fn, graphdef=graphdef, other=other),
+      l2_clip_norm=CLIP_NORM,
+      batch_argnums=(1, 2),  # x and y are batched
+      keep_batch_dim=True,   # Return per-example gradients
```
Comment (ryan112358) [inline examples/dp_sgd_transformer_nnx.py:312]: Delete this inline comment, it's not quite right.  All it means is when computing per-example gradients, pass in batches with a leading axis size of (1,), rather than stripping it out completely.  

Actually you can just delete the whole line since this is the default value

File: examples/dp_sgd_transformer_nnx.py [hunk 16]
```diff
@@ -0,0 +1,390 @@
...
+    return params, opt_state, noise_state, loss.values.mean()
+
+  # Training loop
+  print(f'Training for {NUM_STEPS} steps...')
+  iterator = plan.batch_selection_strategy.batch_iterator(dataset_size)
+  for step, batch_indices in enumerate(iterator):
+    if step >= NUM_STEPS:
+      break
+
+    if len(batch_indices) == 0:
```
Comment (ryan112358) [inline examples/dp_sgd_transformer_nnx.py:372]: Technically you still need to add noise even when you have an empty batch.  If removing this lin breaks something internal to Jax privacy downstream, feel free to leave a TODO for now and we can fix it in follow up.  It it doesn't break anything, then just delete this if -> continue block

File: examples/dp_sgd_transformer_nnx.py [hunk 17]
```diff
@@ -0,0 +1,390 @@
...
+      continue
+
+    # Construct batch from indices
+    x = np.stack([data[i : i + CONTEXT_LENGTH] for i in batch_indices])
```
Comment (ryan112358) [inline examples/dp_sgd_transformer_nnx.py:376]: This logic doesn't look right to me.  For sequence models, data should have shape (N, CONTEXT_LENGTH), and you should be able to do something like data[batch_indices] to get something with shape (B, CONTEXT_LENGTH)

File: examples/dp_sgd_transformer_nnx.py [hunk 18]
```diff
@@ -0,0 +1,390 @@
...
+      num_bands=1,
+      epsilon=EPSILON,
+      delta=DELTA,
+      sampling_prob=BATCH_SIZE / dataset_size,
+  )
+  plan = config.make(grad_fn)
+  privatizer = plan.noise_addition_transform
+  noise_state = privatizer.init(params)
+
+  @jax.jit
```
Comment (ryan112358) [inline examples/dp_sgd_transformer_nnx.py:331]: For efficiency you can donate some argnums (params, opt_state, noise_state)

File: examples/dp_sgd_transformer_nnx.py [hunk 19]
```diff
@@ -0,0 +1,390 @@
...
+    text: The text to encode.
+    stoi: The string-to-index mapping.
+
+  Returns:
+    A numpy array of integer indices.
+  """
+  return np.array([stoi[ch] for ch in text], dtype=np.int32)
+
+
+def get_batch(
```
Comment (ryan112358) [inline examples/dp_sgd_transformer_nnx.py:102]: This function appears unused and can be deleted.

File: examples/user_level_transformer_example.py [hunk 20]
```diff
@@ -0,0 +1,209 @@
...
+    x = nn.Dense(self.vocab_size)(x)
+    return x
+
+
+def get_synthetic_data(
+    num_users: int,
+    num_examples_per_user: list[int],
+    seq_len: int,
+    vocab_size: int,
+):
```
Comment (ryan112358) [inline examples/user_level_transformer_example.py:78]: Can you document the shapes of the output arrays?

File: examples/user_level_transformer_example.py [hunk 21]
```diff
@@ -0,0 +1,209 @@
...
+
+  # Wrap the plan's strategy with UserSelectionStrategy
+  # We assume plan.batch_selection_strategy is compatible
+  # (CyclicPoissonSampling)
+  user_strategy = UserSelectionStrategy(
+      base_strategy=plan.batch_selection_strategy,
+      examples_per_user_per_batch=EXAMPLES_PER_USER,
+  )
+
+  @jax.jit
```
Comment (ryan112358) [inline examples/user_level_transformer_example.py:171]: You could donate_argnums here as well

File: jax_privacy/experimental/execution_plan.py [hunk 22]
```diff
@@ -177,11 +177,13 @@ class BandMFExecutionPlanConfig:
 
   Standard DP-SGD is the special case `num_bands=1`. A typical configuration
   for Poisson-sampled DP-SGD is:
+
```
Comment (ryan112358) [inline jax_privacy/experimental/execution_plan.py:180]: I think this file changed, please rebase to head and revert this change
